File: go/adbc/driver/clickhouse/connection.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package clickhouse

import (
	"context"
	"crypto/tls"
	"fmt"
	"github.com/ClickHouse/clickhouse-go/v2"
	clickhouseDriver "github.com/ClickHouse/clickhouse-go/v2/lib/driver"
	"github.com/apache/arrow-adbc/go/adbc"
	"github.com/apache/arrow-adbc/go/adbc/driver/internal/driverbase"
	"github.com/apache/arrow/go/v17/arrow"
	"strings"
)

type connectionImpl struct {
	driverbase.ConnectionImplBase

	address  []string
	protocol clickhouse.Protocol
	database string
	username string
	password string
	table    string

	resultRecordBufferSize int
	prefetchConcurrency    int

	conn clickhouseDriver.Conn
}

// GetCurrentCatalog implements driverbase.CurrentNamespacer.
func (c *connectionImpl) GetCurrentCatalog() (string, error) {
	return c.database, nil
}

// GetCurrentDbSchema implements driverbase.CurrentNamespacer.
func (c *connectionImpl) GetCurrentDbSchema() (string, error) {
	return c.table, nil
}

// SetCurrentCatalog implements driverbase.CurrentNamespacer.
func (c *connectionImpl) SetCurrentCatalog(value string) error {
	c.database = value
	return nil
}

// SetCurrentDbSchema implements driverbase.CurrentNamespacer.
func (c *connectionImpl) SetCurrentDbSchema(value string) error {
	c.table = value
	return nil
}

// ListTableTypes implements driverbase.TableTypeLister.
func (c *connectionImpl) ListTableTypes(ctx context.Context) ([]string, error) {
	// todo: find clickhouse enums/get from connection for this
	return []string{"BASE TABLE", "TEMPORARY TABLE", "VIEW"}, nil
}

// SetAutocommit implements driverbase.AutocommitSetter.
func (c *connectionImpl) SetAutocommit(enabled bool) error {
	if enabled {
		return nil
	}
	return adbc.Error{
		Code: adbc.StatusNotImplemented,
		Msg:  "SetAutocommit to `false` is not yet implemented",
	}
}

// Commit commits any pending transactions on this connection, it should
// only be used if autocommit is disabled.
//
// Behavior is undefined if this is mixed with SQL transaction statements.
func (c *connectionImpl) Commit(_ context.Context) error {
	return adbc.Error{
		Code: adbc.StatusNotImplemented,
		Msg:  "Commit not yet implemented for BigQuery driver",
	}
}

// Rollback rolls back any pending transactions. Only used if autocommit
// is disabled.
//
// Behavior is undefined if this is mixed with SQL transaction statements.
func (c *connectionImpl) Rollback(_ context.Context) error {
	return adbc.Error{
		Code: adbc.StatusNotImplemented,
		Msg:  "Rollback not yet implemented for BigQuery driver",
	}
}

// Close closes this connection and releases any associated resources.
func (c *connectionImpl) Close() error {
	return c.conn.Close()
}

func (c *connectionImpl) newConnection(ctx context.Context) error {
	var err error
	if c.address == nil || len(c.address) == 0 {
		return adbc.Error{
			Code: adbc.StatusInvalidArgument,
			Msg:  "clickhouse address not given",
		}
	}
	connectionOptions := clickhouse.Options{
		Addr:     c.address,
		Protocol: c.protocol,
		Auth: clickhouse.Auth{
			Database: c.database,
			Username: c.username,
			Password: c.password,
		},
		Compression: &clickhouse.Compression{
			Method: clickhouse.CompressionLZ4,
		},
		Debugf: func(format string, v ...interface{}) {
			fmt.Printf(format, v)
		},
		TLS: &tls.Config{InsecureSkipVerify: true},
	}
	c.conn, err = clickhouse.Open(&connectionOptions)
	if err != nil {
		return adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  err.Error(),
		}
	}
	if err := c.conn.Ping(ctx); err != nil {
		if exception, ok := err.(*clickhouse.Exception); ok {
			return adbc.Error{
				Code: adbc.StatusInternal,
				Msg:  fmt.Sprintf("Exception [%d] %s \n%s\n", exception.Code, exception.Message, exception.StackTrace),
			}
		}
		return adbc.Error{
			Code: adbc.StatusInternal,
			Msg:  err.Error(),
		}
	}
	return err
}

// Metadata methods
// Generally these methods return an array.RecordReader that
// can be consumed to retrieve metadata about the database as Arrow
// data. The returned metadata has an expected schema given in the
// doc strings of the specific methods. Schema fields are nullable
// unless otherwise marked. While no Statement is used in these
// methods, the result set may count as an active statement to the
// driver for the purposes of concurrency management (e.g. if the
// driver has a limit on concurrent active statements and it must
// execute a SQL query internally in order to implement the metadata
// method).
//
// Some methods accept "search pattern" arguments, which are strings
// that can contain the special character "%" to match zero or more
// characters, or "_" to match exactly one character. (See the
// documentation of DatabaseMetaData in JDBC or "Pattern Value Arguments"
// in the ODBC documentation.) Escaping is not currently supported.
// GetObjects gets a hierarchical view of all catalogs, database schemas,
// tables, and columns.
//
// The result is an Arrow Dataset with the following schema:
//
//	Field Name                  | Field Type
//	----------------------------|----------------------------
//	catalog_name                | utf8
//	catalog_db_schemas          | list<DB_SCHEMA_SCHEMA>
//
// DB_SCHEMA_SCHEMA is a Struct with the fields:
//
//	Field Name                  | Field Type
//	----------------------------|----------------------------
//	db_schema_name              | utf8
//	db_schema_tables            | list<TABLE_SCHEMA>
//
// TABLE_SCHEMA is a Struct with the fields:
//
//	Field Name                  | Field Type
//	----------------------------|----------------------------
//	table_name                  | utf8 not null
//	table_type                  | utf8 not null
//	table_columns               | list<COLUMN_SCHEMA>
//	table_constraints           | list<CONSTRAINT_SCHEMA>
//
// COLUMN_SCHEMA is a Struct with the fields:
//
//	Field Name                  | Field Type          | Comments
//	----------------------------|---------------------|---------
//	column_name                 | utf8 not null       |
//	ordinal_position            | int32               | (1)
//	remarks                     | utf8                | (2)
//	xdbc_data_type              | int16               | (3)
//	xdbc_type_name              | utf8                | (3)
//	xdbc_column_size            | int32               | (3)
//	xdbc_decimal_digits         | int16               | (3)
//	xdbc_num_prec_radix         | int16               | (3)
//	xdbc_nullable               | int16               | (3)
//	xdbc_column_def             | utf8                | (3)
//	xdbc_sql_data_type          | int16               | (3)
//	xdbc_datetime_sub           | int16               | (3)
//	xdbc_char_octet_length      | int32               | (3)
//	xdbc_is_nullable            | utf8                | (3)
//	xdbc_scope_catalog          | utf8                | (3)
//	xdbc_scope_schema           | utf8                | (3)
//	xdbc_scope_table            | utf8                | (3)
//	xdbc_is_autoincrement       | bool                | (3)
//	xdbc_is_generatedcolumn     | utf8                | (3)
//
// 1. The column's ordinal position in the table (starting from 1).
// 2. Database-specific description of the column.
// 3. Optional Value. Should be null if not supported by the driver.
//    xdbc_values are meant to provide JDBC/ODBC-compatible metadata
//    in an agnostic manner.
//
// CONSTRAINT_SCHEMA is a Struct with the fields:
//
//	Field Name                  | Field Type          | Comments
//	----------------------------|---------------------|---------
//	constraint_name             | utf8                |
//	constraint_type             | utf8 not null       | (1)
//	constraint_column_names     | list<utf8> not null | (2)
//	constraint_column_usage     | list<USAGE_SCHEMA>  | (3)
//
// 1. One of 'CHECK', 'FOREIGN KEY', 'PRIMARY KEY', or 'UNIQUE'.
// 2. The columns on the current table that are constrained, in order.
// 3. For FOREIGN KEY only, the referenced table and columns.
//
// USAGE_SCHEMA is a Struct with fields:
//
//	Field Name                  | Field Type
//	----------------------------|----------------------------
//	fk_catalog                  | utf8
//	fk_db_schema                | utf8
//	fk_table                    | utf8 not null
//	fk_column_name              | utf8 not null
//
// For the parameters: If nil is passed, then that parameter will not
// be filtered by at all. If an empty string, then only objects without
// that property (ie: catalog or db schema) will be returned.
//
// tableName and columnName must be either nil (do not filter by
// table name or column name) or non-empty.
//
// All non-empty, non-nil strings should be a search pattern (as described
// earlier).

func (c *connectionImpl) GetTableSchema(ctx context.Context, catalog *string, dbSchema *string, tableName string) (*arrow.Schema, error) {
	return c.getTableSchemaWithFilter(ctx, catalog, dbSchema, tableName, nil)
}

// NewStatement initializes a new statement object tied to this connection
func (c *connectionImpl) NewStatement() (adbc.Statement, error) {
	return &statement{
		connectionImpl:         c,
		resultRecordBufferSize: c.resultRecordBufferSize,
		prefetchConcurrency:    c.prefetchConcurrency,
	}, nil
}

func (c *connectionImpl) GetOption(key string) (string, error) {
	switch key {
	case OptionStringProtocol:
		switch c.protocol {
		case clickhouse.HTTP:
			return OptionValueProtocolHTTP, nil
		case clickhouse.Native:
			return OptionValueProtocolNative, nil
		}
	default:
		return c.ConnectionImplBase.GetOption(key)
	}
	return "", nil
}

func (c *connectionImpl) SetOption(key string, value string) error {
	switch key {
	case OptionStringProtocol:
		switch strings.ToLower(value) {
		case OptionValueProtocolNative:
			c.protocol = clickhouse.Native
		case OptionValueProtocolHTTP:
			c.protocol = clickhouse.HTTP
		}
	default:
		return adbc.Error{
			Code: adbc.StatusInvalidArgument,
			Msg:  fmt.Sprintf("unknown statement string type option `%s`", key),
		}
	}
	return nil
}

func (c *connectionImpl) GetOptionInt(key string) (int64, error) {
	switch key {
	case OptionIntQueryResultBufferSize:
		return int64(c.resultRecordBufferSize), nil
	case OptionIntQueryPrefetchConcurrency:
		return int64(c.prefetchConcurrency), nil
	default:
		return c.ConnectionImplBase.GetOptionInt(key)
	}
}

func (c *connectionImpl) SetOptionInt(key string, value int64) error {
	switch key {
	case OptionIntQueryResultBufferSize:
		c.resultRecordBufferSize = int(value)
		return nil
	case OptionIntQueryPrefetchConcurrency:
		c.prefetchConcurrency = int(value)
		return nil
	default:
		return c.ConnectionImplBase.SetOptionInt(key, value)
	}
}

func (c *connectionImpl) getTableSchemaWithFilter(ctx context.Context, catalog *string, dbSchema *string, tableName string, columnName *string) (*arrow.Schema, error) {
	return nil, adbc.Error{
		Code: adbc.StatusNotImplemented,
		Msg:  "getTableSchemaWithFilter is not yet implemented",
	}
}

File: glib/adbc-arrow-glib/connection.c
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <adbc-arrow-glib/connection.h>

/**
 * SECTION: connection
 * @title: GADBCArrowStatement
 * @include: adbc-arrow-glib/adbc-arrow-glib.h
 *
 * #GADBCArrowConnection is a class for Arrow GLib integrated
 * connection.
 */

G_DEFINE_TYPE(GADBCArrowConnection, gadbc_arrow_connection, GADBC_TYPE_CONNECTION)

static void gadbc_arrow_connection_init(GADBCArrowConnection* connection) {}

static void gadbc_arrow_connection_class_init(GADBCArrowConnectionClass* klass) {}

/**
 * gadbc_arrow_connection_new:
 * @error: (nullable): Return location for a #GError or %NULL.
 *
 * Returns: A newly created #GADBCArrowConnection.
 *
 * Since: 1.0.0
 */
GADBCArrowConnection* gadbc_arrow_connection_new(GError** error) {
  GADBCArrowConnection* connection = g_object_new(GADBC_ARROW_TYPE_CONNECTION, NULL);
  if (gadbc_connection_initialize(GADBC_CONNECTION(connection),
                                  "[adbc-arrow][connection][new]", error)) {
    return connection;
  } else {
    g_object_unref(connection);
    return NULL;
  }
}

static GArrowRecordBatchReader* gadbc_arrow_import_array_stream(
    gpointer c_abi_array_stream, GError** error) {
  if (!c_abi_array_stream) {
    return NULL;
  }
  GArrowRecordBatchReader* reader =
      garrow_record_batch_reader_import(c_abi_array_stream, error);
  g_free(c_abi_array_stream);
  return reader;
}

static GArrowSchema* gadbc_arrow_import_schema(gpointer c_abi_schema, GError** error) {
  if (!c_abi_schema) {
    return NULL;
  }
  GArrowSchema* schema = garrow_schema_import(c_abi_schema, error);
  g_free(c_abi_schema);
  return schema;
}

/**
 * gadbc_arrow_connection_get_info:
 * @connection: A #GADBCArrowConnection.
 * @info_codes: (nullable) (array length=n_info_codes): A list of
 *   metadata codes to fetch, or %NULL to fetch all.
 * @n_info_codes: The length of the info_codes parameter. Ignored if
 *   info_codes is %NULL.
 * @error: (nullable): Return location for a #GError or %NULL.
 *
 * This is a wrapper of gadbc_arrow_connection_get_info().
 *
 * Returns: (transfer full) (nullable): The result set as
 *   #GArrowRecordBatchReader on success, %NULL on error.
 *
 * Since: 1.0.0
 */
GArrowRecordBatchReader* gadbc_arrow_connection_get_info(GADBCArrowConnection* connection,
                                                         guint32* info_codes,
                                                         gsize n_info_codes,
                                                         GError** error) {
  gpointer c_abi_array_stream = gadbc_connection_get_info(
      GADBC_CONNECTION(connection), info_codes, n_info_codes, error);
  return gadbc_arrow_import_array_stream(c_abi_array_stream, error);
}

/**
 * gadbc_arrow_connection_get_objects:
 * @connection: A #GADBCArrowConnection.
 * @depth: The level of nesting to display. If
 *   @GADBC_OBJECT_DEPTH_ALL, display all levels. If
 *   @GADBC_OBJECT_DEPTH_CATALOGS, display only catalogs
 *   (i.e. `catalog_schemas` will be null). If
 *   @GADBC_OBJECT_DEPTH_DB_SCHEMAS, display only catalogs and schemas
 *   (i.e. `db_schema_tables` will be null). if
 *   @GADBC_OBJECT_DEPTH_TABLES, display only catalogs, schemas and
 *   tables (i.e. `table_columns` and `table_constraints` will be
 *   null).
 * @catalog: (nullable): Only show tables in the given catalog. If
 *   %NULL, do not filter by catalog. If an empty string, only show
 *   tables without a catalog. May be a search pattern (see section
 *   documentation).
 * @db_schema: (nullable): Only show tables in the given database
 *   schema. If %NULL, do not filter by database schema. If an empty
 *   string, only show tables without a database schema. May be a
 *   search pattern (see section documentation).
 * @table_name: (nullable): Only show tables with the given name. If
 *   %NULL, do not filter by name. May be a search pattern (see
 *   section documentation).
 * @table_types: (nullable) (array zero-terminated=1): Only show
 *   tables matching one of the given table types. If %NULL, show
 *   tables of any type. Valid table types can be fetched from
 *   gadbc_connection_get_table_types(). Terminate the list with a
 *   %NULL entry.
 * @column_name: (nullable): Only show columns with the given name. If
 *   %NULL, do not filter by name. May be a search pattern (see section
 *   documentation).
 * @error: (nullable): Return location for a #GError or %NULL.
 *
 * This is a wrapper of gadbc_connection_get_objects().
 *
 * Returns: (transfer full) (nullable): The result set as
 *   #GArrowRecordBatchReader on success, %NULL on error.
 *
 * Since: 1.0.0
 */
GArrowRecordBatchReader* gadbc_arrow_connection_get_objects(
    GADBCArrowConnection* connection, GADBCObjectDepth depth, const gchar* catalog,
    const gchar* db_schema, const gchar* table_name, const gchar** table_types,
    const gchar* column_name, GError** error) {
  gpointer c_abi_array_stream = gadbc_connection_get_objects(
      GADBC_CONNECTION(connection), depth, catalog, db_schema, table_name, table_types,
      column_name, error);
  return gadbc_arrow_import_array_stream(c_abi_array_stream, error);
}

/**
 * gadbc_arrow_connection_get_table_schema:
 * @connection: A #GADBCArrowConnection.
 * @catalog: (nullable): A catalog or %NULL if not applicable.
 * @db_schema: (nullable): A database schema or %NULL if not applicable.
 * @table_name: A table name.
 * @error: (nullable): Return location for a #GError or %NULL.
 *
 * This is a wrapper of gadbc_connection_get_table_schema().
 *
 * Returns: (transfer full) (nullable): The result set as
 *   #GArrowSchema on success, %NULL on error.
 *
 * Since: 1.0.0
 */
GArrowSchema* gadbc_arrow_connection_get_table_schema(GADBCArrowConnection* connection,
                                                      const gchar* catalog,
                                                      const gchar* db_schema,
                                                      const gchar* table_name,
                                                      GError** error) {
  gpointer c_abi_schema = gadbc_connection_get_table_schema(
      GADBC_CONNECTION(connection), catalog, db_schema, table_name, error);
  return gadbc_arrow_import_schema(c_abi_schema, error);
}

/**
 * gadbc_arrow_connection_get_table_types:
 * @connection: A #GADBCArrowConnection.
 * @error: (nullable): Return location for a #GError or %NULL.
 *
 * This is a wrapper of gadbc_connection_get_table_types().
 *
 * Returns: (transfer full) (nullable): The result set as
 *   #GArrowRecordBatchReader on success, %NULL on error.
 *
 * Since: 1.0.0
 */
GArrowRecordBatchReader* gadbc_arrow_connection_get_table_types(
    GADBCArrowConnection* connection, GError** error) {
  gpointer c_abi_array_stream =
      gadbc_connection_get_table_types(GADBC_CONNECTION(connection), error);
  return gadbc_arrow_import_array_stream(c_abi_array_stream, error);
}

/**
 * gadbc_arrow_connection_get_statistics:
 * @connection: A #GADBCArrowConnection.
 * @catalog: (nullable): A catalog or %NULL if not applicable.
 * @db_schema: (nullable): A database schema or %NULL if not applicable.
 * @table_name: (nullable): A table name.
 * @approximate: Whether approximate values are allowed or not. If
 *   this is %TRUE, best-effort, approximate or cached values may be
 *   returned. Otherwise, exact values are requested. Note that the
 *   database may return approximate values regardless as indicated
 *   in the result. Request exact values may be expensive or
 *   unsupported.
 * @error: (nullable): Return location for a #GError or %NULL.
 *
 * This is a wrapper of gadbc_arrow_connection_get_statistics().
 *
 * Returns: (transfer full) (nullable): The result set as
 *   #GArrowRecordBatchReader on success, %NULL on error.
 *
 * Since: 1.0.0
 */
GArrowRecordBatchReader* gadbc_arrow_connection_get_statistics(
    GADBCArrowConnection* connection, const gchar* catalog, const gchar* db_schema,
    const gchar* table_name, gboolean approximate, GError** error) {
  gpointer c_abi_array_stream = gadbc_connection_get_statistics(
      GADBC_CONNECTION(connection), catalog, db_schema, table_name, approximate, error);
  return gadbc_arrow_import_array_stream(c_abi_array_stream, error);
}

/**
 * gadbc_arrow_connection_get_statistic_names:
 * @connection: A #GADBCArrowConnection.
 * @error: (nullable): Return location for a #GError or %NULL.
 *
 * This is a wrapper of gadbc_connection_get_statistic_names().
 *
 * Returns: (transfer full) (nullable): The result set as
 *   #GArrowRecordBatchReader on success, %NULL on error.
 *
 * Since: 1.0.0
 */
GArrowRecordBatchReader* gadbc_arrow_connection_get_statistic_names(
    GADBCArrowConnection* connection, GError** error) {
  gpointer c_abi_array_stream =
      gadbc_connection_get_statistic_names(GADBC_CONNECTION(connection), error);
  return gadbc_arrow_import_array_stream(c_abi_array_stream, error);
}
